src/codegen/compile.ml

```diff
@@ -9351,8 +9510,371 @@ module FuncDec = struct
+  let async_destabilization_reply_callback env =
+    E.add_fun_ptr env (E.built_in env async_destabilization_reply_callback_name)
+
+  let async_destabilization_reject_callback_name = "@async_destabilization_reject_callback"
+  let async_destabilization_reject_callback env =
+    E.add_fun_ptr env (E.built_in env async_destabilization_reject_callback_name)
+
+  let call_async_destabilization env =
+    IC.get_self_reference env ^^ Blob.as_ptr_len env ^^
+    Blob.lit_ptr_len env async_destabilization_method_name ^^
+    compile_const_32 (async_destabilization_reply_callback env) ^^ compile_const_32 0l ^^
+    compile_const_32 (async_destabilization_reject_callback env) ^^ compile_const_32 0l ^^
+    IC.system_call env "call_new_64" ^^
+    IC.system_call env "call_perform" ^^
+    G.i (Convert (Wasm_exts.Values.I64 I64Op.ExtendUI32)) ^^
+    E.then_trap_with env "Async destabilization increment call failed"
+
+  let complete_graph_destabilization env = 
+    IC.initialize_main_actor env ^^
+    (* Allow other messages and allow garbage collection. *)
+    E.call_import env "rts" "start_gc_after_destabilization" ^^
+    Lifecycle.trans env Lifecycle.Idle
+
+  let define_async_destabilization_reply_callback env =
+    Func.define_built_in env async_destabilization_reply_callback_name ["env", I32Type] [] (fun env ->
+      get_destabilized_actor env ^^
+      compile_test I64Op.Eqz ^^
+      E.if0
+        begin
+          (* Trigger next async destabilization increment. *)
+          call_async_destabilization env
+        end
+        begin
+          (* Send static reply of sucessful async destabilzation sequence. *)
+          IC.static_nullary_reply env
+          (* Stay in lifecycle state `InDestabilization`. *)
+        end)
+  
+  let define_async_destabilization_reject_callback env =
+    Func.define_built_in env async_destabilization_reject_callback_name ["env", I32Type] [] (fun env ->
+      IC.error_message env ^^
+      Blob.as_ptr_len env ^^
+      IC.system_call env "msg_reject_64")
+
+  let destabilization_increment env actor_type =
+    get_destabilized_actor env ^^
+    compile_test I64Op.Eqz ^^
+    (E.if0
+      begin
+        GraphCopyStabilization.graph_destabilization_increment env ^^
+        (E.if0
+          begin
+            GraphCopyStabilization.get_graph_destabilized_actor env actor_type ^^
+            set_destabilized_actor env ^^
+            complete_graph_destabilization env
+          end
+          G.nop)
+      end
+      G.nop)
+
+  let export_async_destabilization_method env actor_type =
+    let name = async_destabilization_method_name in
+    begin match E.mode env with
+    | Flags.ICMode | Flags.RefMode ->
+      Func.define_built_in env name [] [] (fun env ->
+        IC.assert_caller_self_or_controller env ^^
+        (* Stay in lifecycle state `InDestabilization` if not yet completed. *)
+        destabilization_increment env actor_type ^^
+        IC.static_nullary_reply env
```

Comment:
Because this calls IC.static_nullary_reply env, I think the asynchronous call will immediately complete, like oneway message, triggering the destablization to run in the background.

Maybe that's intentional but, if you wanted this to only respond once completed,  I think you only want to issue the reply in `complete_graph_destabilization env`.

Reply:
Thanks for checking. The logic is indeed quite complicated: This is only an internal async helper function which gets called by another outer async function `call_async_destabilization` in an async loop/chain style. The `call_async_destabilization` that is initiated by the external `__motoko_destabilize_after_upgrade`. The outer async function should only return when the entire destabilization has finished.